src/components/auth/authenticator.ts

```diff
@@ -0,0 +1,113 @@
+import debug from 'debug';
+import EventEmitter from 'eventemitter3';
+import Keycloak_, { KeycloakInstance } from 'keycloak-js';
+import { User } from './user';
+
+export type AuthenticatorEventType = {
+  userChanged: [User | null];
+};
+
+export default interface Authenticator extends EventEmitter<AuthenticatorEventType> {
+  readonly user?: User;
+  readonly sossToken?: string;
+
+  /**
+   * It is required to call this before using any of the authenticator functions.
+   */
+  init(): Promise<void>;
+
+  /**
+   * Note: This redirects to external login page so it will never return.
+   */
+  login(): Promise<never>;
+
+  /**
+   * Note: This redirects to external logout page so it will never return.
+   */
+  logout(): Promise<void>;
+}
+
+export type AuthConfig<T = typeof Keycloak_> = T extends (config: infer U) => unknown ? U : never;
+
+export class DefaultAuthenticator extends EventEmitter<AuthenticatorEventType>
+  implements Authenticator {
+  get user(): User | undefined {
+    return this._user;
+  }
+
+  get sossToken(): string | undefined {
+    return this._inst.idToken;
+  }
+
+  constructor(config: AuthConfig, redirectUri?: string) {
+    super();
+    this._inst = Keycloak_(config);
+    this._redirectUri = redirectUri;
+  }
+
+  async init() {
+    debug.log('initializing authenticator');
+
+    this._inst.onAuthSuccess = async () => {
+      this._user = {
+        username: (this._inst.idTokenParsed as any).preferred_username,
+      };
+      debug.log('authenticated as', this._user.username);
+      this.emit('userChanged', this._user);
+    };
+
+    this._inst.onAuthLogout = () => {
+      debug.log('logout');
+      this._user = undefined;
+      this.emit('userChanged', null);
+    };
+
+    const token = getLocalStorage('token');
+    const idToken = getLocalStorage('idToken');
+    const refreshToken = getLocalStorage('refreshToken');
+    await this._inst.init({ redirectUri: this._redirectUri, token, idToken, refreshToken });
+    try {
+      const refreshed = await this._inst.updateToken(30);
+      refreshed && debug.log('token refreshed');
+    } catch {}
+    setOrClearLocalStorage('token', this._inst.token);
```

Comment:
Should be worried about the security of setting the token on the local storage?

Reply:
afaik this should be fine. The tokens are signed with RSA and encodes various information like expiry time etc, they are meant to be stored on the client side so we don't need to repeated connect to the auth server. This is what I understand from the oauth/openid connect guides, security is a fickle thing so do let me know if there is a problem with this.

Reply:
Yeah, I was thinking that maybe it'll be more secure (if possible) to store the `token` on memory. I left a couple of short articles about the subject: 
https://auth0.com/docs/tokens/token-storage#browser-in-memory-scenarios 
https://auth0.com/docs/tokens/token-storage#browser-local-storage-scenarios

Reply:
Thats a good point, I think it's still "ok" security wise as long as we protect against XSS. But I don't mind using in-memory tokens, storing the token in a persistent storage has a small downside in sso that if the user logs in from romi-dashboard -> logs out from another app -> comes back to romi-dashboard, the access token is no longer valid but romi-dashboard doesn't know about it.